internals/controllers/subnamespace/subnamespace_init.go

```diff
@@ -0,0 +1,219 @@
+		return ctrl.Result{}, fmt.Errorf("failed to append for subnamespace '%s': "+err.Error(), snsName)
+	}
+	logger.Info("successfully appended annotations for subnamespace", "subnamespace", snsObject.Object.GetName())
+
+	if err := r.ensureSNSInDB(ctx, snsObject); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure presence in namespaceDB for subnamespace '%s'", snsObject.Object.GetName())
+	}
+	logger.Info("successfully ensured presence in namespaceDB for subnamespace", "subnamespace", snsObject.Object.GetName())
+
+	err = snsObject.UpdateObject(func(object client.Object, log logr.Logger) (client.Object, logr.Logger) {
+		object.(*danav1.Subnamespace).Status.Phase = danav1.Created
+		log = log.WithValues("updated subnamespace phase", danav1.Created)
+		return object, log
+	})
+
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set status '%s' for subnamespace '%s'", danav1.Created, snsName)
+	}
+	logger.Info("successfully set status for subnamespace", "phase", danav1.Created, "subnamespace", snsName)
+
+	return ctrl.Result{}, nil
+}
+
+// ensureSNSNamespace makes sure a namespace is created for subnamespace
+func ensureSNSNamespace(snsParentNS, snsObject *utils.ObjectContext) error {
+	snsNamespace := composeSNSNamespace(snsParentNS, snsObject)
+	snsName := snsObject.Object.GetName()
+
+	snsNSObject, err := utils.NewObjectContext(snsObject.Ctx, snsObject.Client, types.NamespacedName{Name: snsName}, snsNamespace)
+	if err != nil {
+		return err
+	}
+
+	if snsNSObject.IsPresent() {
+		annots := snsNamespace.GetAnnotations()
+		if err := snsNSObject.AppendAnnotations(annots); err != nil {
+			return err
+		}
+
+		labels := snsNamespace.GetLabels()
+		if err := snsNSObject.AppendLabels(labels); err != nil {
+			return err
+		}
+	}
+
+	if err := snsNSObject.EnsureCreateObject(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// composeSNSNamespace creates a new namespace for a particular subnamespace with labels
+// and annotations based on the namespace linked to the parent of the subnamespace
+func composeSNSNamespace(snsParentNS, snsObject *utils.ObjectContext) *corev1.Namespace {
+	nsName := snsObject.Object.GetName()
+	labels, annots := utils.GetNSLabelsAnnotationsBasedOnParent(snsParentNS, nsName)
+
+	// add the ResourcePool label separately from the function
+	labels[danav1.ResourcePool] = snsObject.Object.GetLabels()[danav1.ResourcePool]
+
+	return ComposeNamespace(nsName, labels, annots)
+}
+
+// ensureDefaultSNSResourceQuota creates a ResourceQuota object with some default values
+// that we would like to limit and are not set by the user
+func ensureDefaultSNSResourceQuota(snsObject *utils.ObjectContext) error {
+	snsName := snsObject.Object.GetName()
+	composedDefaultRQ := ComposeResourceQuota(snsName, snsName, danav1.DefaultQuotaHard)
+
+	childResourceQuota, err := utils.NewObjectContext(snsObject.Ctx, snsObject.Client, types.NamespacedName{Name: snsName, Namespace: snsName}, composedDefaultRQ)
```

Comment:
Can't do:

```
	if snsDefaultRQ, err := utils.NewObjectContext(snsObject.Ctx, snsObject.Client, types.NamespacedName{Name: snsName, Namespace: snsName}, composedDefaultRQ); err != nil {
		return err
	}
```

Since I need to use the `snsDefaultRQ` (new name) outside the scope of the `if`